### PR TITLE
LEAF-4349 - Query step actions

### PR DESCRIPTION
**Pull request by mgaoVA**
This expands `api/form/query`: Find all records if a specific action was taken for a given step.

No additional DB indexes seem to make a meaningful difference in query performance. Observed up to 0.14s query time on a site containing 2M actions.

Automated tests pending merge of PR#2390

### This PR is dependent on:
- https://github.com/department-of-veterans-affairs/LEAF/pull/2395
- https://github.com/department-of-veterans-affairs/LEAF/pull/2390

### Potential Impact
Minimal because this adds new functionality without touching existing core logic.

### Testing
- [ ] Passes automated tests
- [ ] Create a record and apply actions to it. Using the Homepage Advanced Search or Report Builder, use the new `Record Actions ...` term to find the record by using matching Step and Actions.

### CODE REVIEW

**Review by jampaul3 — APPROVED** [review]

**Review by shaneodd — APPROVED** [review]

**Review by Pelentan — CHANGES_REQUESTED** [review]

File: LEAF_Request_Portal/js/formSearch.js [hunk 1]
```diff
@@ -685,10 +691,54 @@ var LeafFormSearch = function (containerID) {
...
+            }
+            options += '</select>';
+        }
+
+        return options;
+    }
+
+    // getWorkflowStepActions retrieves data to support stepAction queries
+    async function getWorkflowStepActions(workflowID) {
+        url =
```
[the remark this answers is not in the document]
Comment (mgaoVA) [reply LEAF_Request_Portal/js/formSearch.js:712]: Yeah this is pretty ugly. I hope I'm not the one who originally wrote it :P

File: LEAF_Request_Portal/js/formSearch.js [hunk 2]
```diff
@@ -1456,6 +1494,76 @@ var LeafFormSearch = function (containerID) {
...
+                        </select>`);
+                    $("#" + prefixID + "widgetMatch_" + widgetID).html('');
+                    if(rootURL == "") {
+                        url = "./api/workflow/steps?x-filterData=workflowID,stepID,stepTitle,description";
+                    } else {
+                        url = rootURL + "api/workflow/steps?x-filterData=workflowID,stepID,stepTitle,description";
+                    }
+
+                    if(cache['api/workflow/steps'] == undefined) {
+                        cache['api/workflow/steps'] = $.ajax({
```
[the remark this answers is not in the document]
Comment (mgaoVA) [reply LEAF_Request_Portal/js/formSearch.js:1510]: Updated

File: LEAF_Request_Portal/sources/Form.php [hunk 3]
```diff
@@ -3397,6 +3397,59 @@ public function query(string $inQuery): mixed
...
+                    {
+                        return 0;
+                    }
+
+                    switch($operator) {
+                        case "=":
+                            $vars[':indicatorID' . $count] = $q['indicatorID']; // this is the stepID
+                            // This checks if someone has taken a specific action for a stepID
+                            // OUTER JOIN gets the most recent action for a specific stepID, since there can be
+                            // loops within a workflow, and people can take different actions later
```
[the remark this answers is not in the document]
Comment (mgaoVA) [reply LEAF_Request_Portal/sources/Form.php:3412]: I thought that too when I saw this technique, but it's actually pretty fast: `Observed up to 0.14s query time on a site containing 2M actions.`, especially considering it's a tough query and we don't need to touch the schema.

In this case it's faster to keep all the processing in the DB. Using the above test, it lets PHP avoid having to process about 75% of the records, and avoid overhead from additional queries.

**Review by jampaul3 — APPROVED** [review]

**Review by Pelentan — APPROVED** [review]

**Comment by mgaoVA** [comment]
@Pelentan I rebased this onto Sprint-75-c1 + https://github.com/department-of-veterans-affairs/LEAF/pull/2419, and added an automated test.

**Pull request merged**